raft-tech/TANF-app
# Issue 430: Add Submit Files button to upload files

### Summary of Changes

This PR adds a Submit Files button to upload file page, improves client-side file validation, and improves upon the the submission UX with alerts and new errors.

![430](https://user-images.githubusercontent.com/12901172/114921258-5c3ec580-9df8-11eb-8943-b309a4e34b50.gif)

Addresses issue #430
_Acceptance criteria as stated below_

### How to Test
_List the steps to test the PR_

1. Log in as any user.
2. Click the Data Files tab.
3. Select a year in the Fiscal Year dropdown.
4. Click Search.
6. Drag a text file to any of the four sections.
7. Check that there were no errors, and the input area changes to display a file icon with the name of the uploaded file.
8. Click Submit Data Files.
9. Ensure a success banner appears with a description of the uploaded sections.
10. Try uploading a non-text file (i.e. a `.pdf` file) and ensure the input changes to an error state.
11. To test the final AC, see [this issue](https://github.com/raft-tech/TANF-app/issues/826) for wiring the frontend to a backend endpoint

**Note:** In order to unblock this PR we are using a stubbed endpoint, so the files, at this point aren't actually going anywhere.

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

    - [x] There is a header with text "[The Selected Year]"
    - [x] There are 4 sections for uploading documents
    - [x] `Submit Data Files` button is added to the page (updated based on final #427)
    - [x] The File Upload inputs accepts files that are not binary
    - [x] Frontend matches #427 delivered and accepted mockup 
    - [x] Files are sent to endpoint when clicking `Submit Data Files` button
    - [x] Test coverage exceeds 90%
    - [x] Create new issue for wiring stubbed endpoints to backend


As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: 95% (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

> NOTE: until we have a proper staging environment this may not be satisfiable prior to merging


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

  - No issues found.
